homeassistant/components/backup/config.py

```diff
@@ -271,16 +338,81 @@ def _delete_filter(
                 """Return backups older than days to delete."""
                 # we need to check here since we await before
                 # this filter is applied
-                if self.days is None:
+                agents_retention = {
+                    agent_id: agent_config.retention
+                    for agent_id, agent_config in manager.config.data.agents.items()
+                }
+                has_agents_retention = any(
+                    agent_retention for agent_retention in agents_retention.values()
+                )
+                has_agents_retention_days = any(
+                    agent_retention and agent_retention.days is not None
+                    for agent_retention in agents_retention.values()
+                )
+                if (global_days := self.days) is None and not has_agents_retention_days:
+                    # No global retention days and no agent retention days
                     return {}
+
                 now = dt_util.utcnow()
-                return {
-                    backup_id: backup
-                    for backup_id, backup in backups.items()
-                    if dt_util.parse_datetime(backup.date, raise_on_error=True)
-                    + timedelta(days=self.days)
-                    < now
-                }
+                if global_days is not None and not has_agents_retention:
+                    # Return early to avoid the longer filtering below.
+                    return {
+                        backup_id: backup
+                        for backup_id, backup in backups.items()
+                        if dt_util.parse_datetime(backup.date, raise_on_error=True)
+                        + timedelta(days=global_days)
+                        < now
+                    }
```

Comment:
This does not run often, I'm not sure it's motivated to keep this special case considering the increased code complexity.

Reply:
Yeah, we can remove it.

Reply:
We need to update many tests if we remove this since we need to replace a MagicMock with a real dataclass. I did that for the new tests. Can we do this in a follow up?

Reply:
https://github.com/home-assistant/core/pull/143821